apollo-federation/src/sources/connect/models.rs

```diff
@@ -91,72 +105,122 @@ impl Connector {
 
         connect_arguments
             .into_iter()
-            .map(move |args| {
-                let source = if let Some(source_name) = args.source {
-                    source_arguments
-                        .iter()
-                        .find(|source| source.name == source_name)
-                } else {
-                    None
-                };
+            .map(|args| Self::from_directives(schema, subgraph_name, spec, args, &source_arguments))
+            .collect()
+    }
 
-                let source_name = source.map(|s| s.name.clone());
-                let connect_http = args.http.expect("@connect http missing");
-                let source_http = source.map(|s| &s.http);
-
-                let transport = HttpJsonTransport::from_directive(connect_http, source_http)?;
-
-                let parent_type_name = args.position.field.type_name().clone();
-                let schema_def = &schema.schema().schema_definition;
-                let on_query = schema_def
-                    .query
-                    .as_ref()
-                    .map(|ty| ty.name == parent_type_name)
-                    .unwrap_or(false);
-                let on_mutation = schema_def
-                    .mutation
-                    .as_ref()
-                    .map(|ty| ty.name == parent_type_name)
-                    .unwrap_or(false);
-                let on_root_type = on_query || on_mutation;
-
-                let id = ConnectId {
-                    label: make_label(subgraph_name, &source_name, &transport),
-                    subgraph_name: subgraph_name.to_string(),
-                    source_name: source_name.clone(),
-                    directive: args.position,
-                };
+    fn from_directives(
+        schema: &Schema,
+        subgraph_name: &str,
+        spec: ConnectSpec,
+        connect: ConnectDirectiveArguments,
+        source_arguments: &[SourceDirectiveArguments],
+    ) -> Result<(ConnectId, Self), FederationError> {
+        let source = if let Some(source_name) = connect.source {
+            source_arguments
+                .iter()
+                .find(|source| source.name == source_name)
+        } else {
+            None
+        };
```

Comment:
Small nit: might be nice to use `Option::and_then` here:

```suggestion
        let source = connect.source.as_ref()
            .and_then(|name| source_arguments.iter()
                .find(|s| s.name == name));
```